### PR TITLE
Add resource pin assignment UI to Servos tab

### PR DESCRIPTION
**Pull request by Manwe-777**
## Summary
Add UI to view and edit motor/servo pin assignments in the Servos tab.

## Changes
- Add resource assignment section with editable dropdowns for motor/servo pins
- Add MSP2_MOTOR_SERVO_RESOURCE support in MSPHelper
- Add FileSystem.js fallback for browsers without File System Access API

## Requires
- Firmware PR: https://github.com/betaflight/betaflight/pull/14943

## Testing
Tested on SPEEDYBEEF405WING with custom firmware build.

## Summary by CodeRabbit

* **New Features**
  * Added a "Resource Assignments" panel in the Servos tab with a responsive two-column layout to view and manage Motor and Servo pin assignments. Dropdowns allow reassignment; edits take effect immediately and enable Save (Save + reboot required to persist). Panel is hidden when firmware resource data is not available.

* **Localization**
  * Added English UI labels, status messages and two hint texts (including CLI examples) for the resource assignment workflow.

[Review Change Stack](https://app.coderabbit.ai/change-stack/betaflight/betaflight-configurator/pull/4881?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by haslinghuis** [comment]
image

**Review by haslinghuis — APPROVED** [review]

File: src/js/msp/MSPHelper.js [hunk 1]
```diff
@@ -2579,6 +2622,40 @@ MspHelper.prototype.sendServoConfigurations = function (onCompleteCallback) {
...
+            .push8(rule.min)
+            .push8(rule.max)
+            .push8(rule.box);
+
+        ruleIndex++;
+        if (ruleIndex === FC.SERVO_RULES.length) {
+            nextFunction = onCompleteCallback;
+        }
+
+        MSP.send_message(MSPCodes.MSP_SET_SERVO_MIX_RULE, buffer, false, nextFunction);
```
Comment (haslinghuis) [inline src/js/msp/MSPHelper.js]: `MSP_SET_SERVO_MIX_RULE` is unused ?

Comment (Manwe-777) [reply src/js/msp/MSPHelper.js]: Just checked again; it was part of the mixers editing work that was introduced on the firmware PR side of this too, but not implemented on the configurator for scope. removing it from this one.

**Comment by Manwe-777** [comment]
Updated the style and fixed the toggle on the page;
image

File: src/components/tabs/ServosTab.vue [hunk 2]
```diff
@@ -276,6 +383,62 @@ export default defineComponent({
...
+            resources[index].pin = newPin;
+            resources[index].ioTag = ioTag;
+            resourcesModified.value = true;
+
+            // Send to FC
+            const label = resourceType === 0 ? "Motor" : "Servo";
+            mspHelper.setMotorServoResource(resourceType, index, ioTag, () => {
+                console.log(`${label} ${index + 1} pin set to ${newPin}`);
+            });
+        }
```
[the remark this answers is not in the document]
Comment (haslinghuis) [reply src/components/tabs/ServosTab.vue]: @coderabbitai instead of updating local state immediately - and restoring on failure - is it not more efficient to only update in the send to FC part of the code ?

**Comment by haslinghuis** [comment]
@Manwe-777 we have motor reordering in motors tab already - was thinking to split the MSP message for MOTOR and SERVO to be separate and perhaps move motor pin assignments to motor component.

@limonspb what do you think ?

**Comment by sugaarK** [comment]
@limonspb ?

**Comment by limonspb** [comment]
oh sorry :)
I like that.
THe only problem i see is that this also would require timers reassignment, preferably automatic... Otherwise users will assign servos, and they won't work in most of FCs...
@Bskimp what you think?

**Comment by Bskimp** [comment]
Hey @Manwe-777 - really nice work here. This fills a real gap, and the paired firmware MSP commands are exactly the kind of foundation the configurator has been missing for resource assignment work.

For context: we have been building a similar Resource Assignments flow on the wing fork ([`Bskimp/betaflight-configurator:wing-main`](https://github.com/Bskimp/betaflight-configurator/tree/wing-main)) for the last few weeks. I had not seen this PR when we started, or I probably would have tried to contribute here directly. We ended up solving a few adjacent problems through bench testing, especially around timer conflicts and automatic pad selection, so I wanted to share the pieces that may be useful.

This is not meant as "please consume our implementation." I think this PR provides the better mainline transport layer; our work is more of a safety/UX layer that could sit on top of it.

The short version, especially in response to the timer-reassignment concern: the UI really wants board-derived timer/DMA data, then the dropdown can rank/select pads based on what the assignment costs instead of showing a flat pin list.

<details>
<summary>Timer and DMA warnings</summary>

### 1. Timer conflicts: a SERVO sharing a TIMx with a MOTOR will fight DSHOT

If the user picks a pin whose timer is already driving a motor, the servo's PWM frequency collides with DSHOT timing. It is obvious on the bench/in the air, but not visible in the current UI.

- **Wing-fork:** [`wingResourceAnalyzer.js#L108-L131`](https://github.com/Bskimp/betaflight-configurator/blob/wing-main/src/js/utils/wingResourceAnalyzer.js#L108-L131) - `deriveWarnings()` emits `code: "servo_on_motor_timer"` whenever a servo's timer appears in the motor-timer set.
- **Injection point in this PR:** below the `servosResourceAssignments` title block in `ServosTab.vue` - render a warnings panel. The data needs a `timer show` parse, or eventually a small MSPv2 extension on top of betaflight/betaflight#14943 that returns `{index, ioTag, timer, dmaStream}` instead of just `{index, ioTag}`.

### 2. DMA awareness: a motor without a DMA stream silently bit-bangs bidir DSHOT

Motors that cannot get a DMA stream and do not have TIMUP burst still "work"; they just fall back to bit-banging, which can hurt RPM telemetry and add jitter. Pilots do not get a warning today.

- **Wing-fork:** [`wingResourceAnalyzer.js#L111-L119`](https://github.com/Bskimp/betaflight-configurator/blob/wing-main/src/js/utils/wingResourceAnalyzer.js#L111-L119) - `motor_no_dma` warning. We also flag `freeDmaStreams.length < 3` as DMA scarcity.
- **Injection point:** same warnings panel as #1. Could also add a per-row badge in the Motor Pins table when `dmaStream === null`. The `MSP2_MOTOR_SERVO_RESOURCE` payload is the right place to start carrying this; adding two more bytes per resource for `(timer, dmaStream)` would close the loop without a second round-trip.

</details>

<details>
<summary>Pin-pool and dropdown behavior</summary>

### 3. The pin pool should come from the connected board

The current `initialPins + assignedPins` approach solves the "vanishing pin" problem for swaps/reverts, but it still only knows about pins that were already assigned. The harder problem is discovering unassigned timer-capable pads without offering pins that are invalid for the connected MCU/target. The connected FC already knows the answer.

- **Wing-fork:** [`wingResourceAnalyzer.js#L39-L62`](https://github.com/Bskimp/betaflight-configurator/blob/wing-main/src/js/utils/wingResourceAnalyzer.js#L39-L62) - `buildTimerLookup` / `buildDmaLookup` join `resource show` + `timer show` + `dma show` keyed on `(peripheral, index)` to produce the actual per-board topology. `freePads` is then "pads with a timer that aren't claimed by a hardware-fixed peripheral."
- **Injection point:** the `availablePins` computed in `ServosTab.vue`. Replace/augment the assigned-pin-only pool with analyzer output for the connected FC. If you want to stay MSP-only, the firmware-side equivalent is roughly "iterate `timerHardware[]`, exclude entries where the IO is owned by a non-MOTOR/SERVO/LED peripheral, return the IO tags."

### 4. The dropdown should know what each option costs

A flat `["A08","A09","B00",...]` list lets a pilot pick a pin currently bound to MOTOR 1 with no warning. Their motor is now broken and they do not know why until reboot or bench testing.

- **Wing-fork:** [`wingRemapRecommender.js#L55-L240`](https://github.com/Bskimp/betaflight-configurator/blob/wing-main/src/js/utils/wingRemapRecommender.js#L55-L240) - `candidatePadsForSlot()` returns `{pad, timer, channel, dmaStream, source, requiresRelease, sharesTimerWithMotor}` per option. Ranking is: zero-churn current pad -> motor-release candidate (with the `resource MOTOR N NONE` line in `requiresRelease`) -> free non-conflict -> free-with-timer-conflict -> LED_STRIP -> UART. Anything currently claimed by something we are not releasing in this batch is filtered out entirely.
- **Injection point:** `availablePins` again, but the dropdown render also needs richer option data. Render `{pad, source, requiresRelease}` so the option text can show e.g. `"A08 - currently MOTOR 1 (will be released)"` and reorder/disable based on safety.

</details>

<details>
<summary>Apply model and testability</summary>

### 5. The live-apply model bites in subtle ways; staged + Save is more forgiving

CodeRabbit already flagged the optimistic-UI desync on `onResourcePinChange`. The deeper problem is that even with success-handling, partway through editing four servos the FC is in a half-reconfigured state. If the user picks a bad pin midway and aborts, they have no way back to "what was loaded when the tab opened" without a reboot.

- **Wing-fork:** [`WingTuningTab.vue#L3190`](https://github.com/Bskimp/betaflight-configurator/blob/wing-main/src/components/tabs/WingTuningTab.vue#L3190) - our save gate is `pinAssignmentDirty.value || mixerDirty.value || motorCountDirty.value`. Pin Assignment edits stage into a `padOverrides` Map (Vue reactive); the CLI batch fires once on Save. We learned this the hard way: Flying Wing + diff-thrust silently saved `smix`/`yaw_type` but not `mmix` on a bench because only one dirty flag was gated. Live-apply makes that class of bug much harder to reason about because the FC state at any moment matches the cumulative trail of dropdown changes instead of one explicit user intent.
- **Injection point:** `onResourcePinChange` in `ServosTab.vue` - keep a `pendingResourceChanges = new Map<key, ioTag>` (key = `${resourceType}:${index}`), update only the local reactive state on dropdown change, and write all entries via `setMotorServoResource` from the existing Save handler. The existing `resourcesModified` ref naturally becomes `pendingResourceChanges.size > 0`.

### 6. Move the analysis logic out of the .vue so it is testable

Almost everything above is pure data-in / data-out, which makes it a good Vitest target. Once it lives inside `ServosTab.vue` reactive state, it becomes much harder to lock down with regression tests.

- **Wing-fork:** [`test/js/utils/wingRemapRecommender.test.js`](https://github.com/Bskimp/betaflight-configurator/blob/wing-main/test/js/utils/wingRemapRecommender.test.js) - 40+ cases covering null-safety, zero-churn, motor-timer partitioning, release-hint generation, silkscreen-vs-free ranking, and related edge cases. Plus [`wingResourceAnalyzer.test.js`](https://github.com/Bskimp/betaflight-configurator/blob/wing-main/test/js/utils/wingResourceAnalyzer.test.js) with golden CLI fixtures and warning coverage.
- **Injection point:** factor `availablePins` + eventually the warning derivation into `src/js/utils/resourceAnalyzer.js` (vendor-neutral name). Pure function `analyzeResources({ motorResources, servoResources, timerData, dmaData }) -> { availablePins, warnings }`. As a bonus, round-trip tests on `pinToIoTag` / `ioTagToPin` would catch bounds issues before they can ship.

</details>

I do not have a strong opinion on whether MOTOR rows should stay in this Servos tab section or move toward the Motors tab, since haslinghuis mentioned that earlier. The analyzer/ranker can feed either UI; the important part is that the pin options are board-derived and timer-aware.

Happy to extract a vendor-neutral cut of the analyzer + candidate-ranker as a follow-up PR layered on top of this one if any of it sounds useful. It would not need to replace the MSP work here. Or if it is easier, I can open a separate tracking issue and we can iterate there.

Either way: thanks for pushing this forward. The firmware MSP commands alone unlock a lot.

**Comment by Bskimp** [comment]
So here's a prime example of why I say the configurator needs to know what the boards pads actually are as far as pin/timer, it needs to be 'smart' about its remapping. 

if you have prior resources off, they disappear from drop down.  For instances - 
image

image
While in cli this would seem normal, but it leaves the users at a point in a gui where they have to apply defaults or restore resources in another way like reset.  It can leave you confused sometimes when plugging in a board that has already had changes to its resources esp if something has been remapped, even worse if you remapped a timer or uart too since some timers are actually on uarts as well.  



Prime example on target I've dealt with extensively lately;
PC6 / UART6_TX on the SpeedyBee F405 Wing 

Pin | AF2 | AF3 | AF8

PC6 | TIM3_CH1 | TIM8_CH1 | USART6_TX
PC7 | TIM3_CH2 | TIM8_CH2 | USART6_RX

Three peripherals sharing two pins. The board has to pick: UART for telemetry/GPS/blackbox, or a TIM3 motor/servo output, or a TIM8 motor/servo output — and the DMA stream available depends on which one you pick (TIM3_CH1 is DMA1 S4 Ch5, TIM8_CH1 is DMA2 S2 Ch7, USART6_TX is DMA2 S6 Ch5 

This is what my resource table ended up looking like in the end 
image
It picks up the stock resources and banks them there, so you always have an idea of what was what just by looking and saying oh, m1 pad use to be b00 timer 3 ch 3, that's where my servo is now.  This also allowed for being able to see/adjust alternate timers for the same pad to keep dshot happy with servos.  The main difference is I built this around being able to 1 click a preset for airplanes, vtail or flying wing so I needed it intelligent enough to handle that, all the nice benefits to users just kinda came along for the ride. 

I tried to do something similar with the resource tables you laid out and made this; but it needs pad awareness still, this is why I say the biggest things needed are the board derived pins and timers and dma. All of them affect each other 
image

Just sharing some pain points I banged my head against the wall with for a LONG WHILE a few weeks ago lol.  It all seemed so simple till I started applying the presets and wondering what in the hell was even going on and hit one problem after another with motors/servos/timers landing in strange places or pads that don't even exist on the board, cuz that's another big one.  Tons of boards will list m1 to m8 while not actually having pads for 5-8, but those timers and pads will show a lot of times as still in the optional resources, you just cant access them without turning off a uart or something else possibly.

**Comment by Bskimp** [comment]
**TL;DR**: The deeper issue isn't the resource UI itself — it's that the configurator can't see a board's *defaulted* resource/timer/DMA state, which leaves partial custom resources, and remap headaches without a ground truth. Three possible fixes outlined below.

Wanted to surface a broader architectural point from extending this work — I added the timer-aware filtering for the servo/motor pin tables on top of the original, and the limitation that keeps coming back isn't really a UI problem, it's that the configurator has no way to know a board's *defaulted* resource/timer/DMA state. That gap drives a whole class of issues the resource UI alone can't fix.

With defaulted-state awareness, the dropdown can show what each pin would release or take over relative to defaults — like this:

image

image


Without it, the same dropdown when motors/servos are set to NONE looks like this — you can still pick pins, but you've lost the "where things used to be" context that makes the labels useful:

image


image

And once you start entering with custom-set resources, the problem compounds. This is my own current planes setup:

image
image



Pads still in use render fine, but unused pads lose their labels. Same root cause — the configurator can only see what the board reports right now, not what it would default to.

After ~6 weeks pushing this around for plane presets/setup, the three concrete paths I see:

1. **CLI** — a `defaults --dryrun` (or similar) that emits the resource/timer/DMA state to stdout without applying it. Cleanest from a firmware standpoint.
2. **Config parsing** — parse `unified_targets/configs/*.config` configurator-side. Works without firmware changes but is slow and lags every time target configs shift.
3. **MSP** — a new MSP that returns the would-be-defaulted resource/timer/DMA map for the current target. Most invasive but the most useful long-term, and the only option that keeps the configurator authoritative.

Without known defaulted values available at any point in the session, there's always going to be a class of weird edges the resource UI can't account for —  partial custom resources, restoring a certain pad after a remap, etc. Worth treating as a separate workstream regardless of where this PR ends up.

Happy to take this on as a separate PR — would want to align on which of the three paths makes most sense before starting. This is also foundational for the wing setup integration I've been working on for the configurator, so the work overlaps either way.

**Comment by Manwe-777** [comment]
Hey there @Bskimp ! Gave your suggestions a pass on another branch, leveraging your wing fork + my MSP commands and it seems to work (also the UI looks way nicer), buut you were right, the culprit is always the defaults + going back to the previous state; Betaflight needs to provide the old state and the available pins in order to remap properly, and have the user know whats available. Im not 100% sure if that is something that we can do easily on the firmware though, as that requires either hacking in the defaults or mapping all boards default resources, like you mentioned. 

https://github.com/Manwe-777/betaflight-configurator/tree/feature/pad-aware-resource-ui
https://github.com/Manwe-777/betaflight/compare/feature/msp-motor-servo-resource...feature/msp-motor-servo-resource-extended

What Claude says;
Practical recommendation
- Short term: Keep MSP + analyzer + recommender + staged saves; treat localStorage as optional UX sugar with clear copy (“not factory default”).
- Medium term: Pick one authority for defaults — ideally firmware MSP or generated configurator data from the same repo as targets — and treat Bskimp’s UI sketch (delta from default) as the north star.
- If you never add ground truth: the UI remains power user + heuristic; acceptable for some audiences, frustrating for wing/board designers and for support.

**Comment by Bskimp** [comment]
big update on the resources not showing when something = none,  I think I finally figured out how to handle the disappearing resources. Decided to try and prototype the unified target parsing and it actually is working quite well! -

image
image
image
image

Thanks again @Manwe-777 for pushing this, it was exactly what was missing from the configurator for resources reampping and was one of the final pieces of the puzzle to get this thing off the ground for an easier set up for BF fixed wing in configurator, let alone just the resource/servo setup aspect alone, awesome job :) Also much appreciate the UI compliment, I am not very confident in making UI look great haha, glad to hear it looks good though

**Review by blckmn — APPROVED** [review]
Should get the CODERABBIT issues sorted please

**Comment by Manwe-777** [comment]
@coderabbitai full review

All previously requested changes have been addressed in subsequent commits, and the latest SonarCloud-only nits (S6638/S6639/S7773 on the new pin helpers, and S3776 cognitive complexity on loadResourceData) have also been resolved on the current HEAD (`cba7087e`). Please run a fresh full review against the new HEAD so the stale CHANGES_REQUESTED state can be updated. Thanks!

File: src/components/tabs/ServosTab.vue [hunk 3]
```diff
@@ -94,45 +94,111 @@
...
+                                    orientation="vertical"
+                                    inverted
+                                    :model-value="getBarHeight(servoData[i - 1] ?? 1500)"
+                                    :max="100"
+                                    color="warning"
+                                    size="2xl"
+                                    :ui="{
+                                        root: '!w-full',
+                                        base: '!w-full !rounded-md border border-(--ui-border)',
+                                        indicator: '!rounded-none !transition-none opacity-(--bar-opacity)',
```
[the remark this answers is not in the document]
Comment (Manwe-777) [reply src/components/tabs/ServosTab.vue]: Applied the Tailwind v4 trailing-`!` form for the three `UProgress` `:ui` overrides in ServosTab.vue (`w-full!`, `rounded-md!`, `transition-none!`) in commit `5c915065`.

I intentionally did **not** touch the identical pattern in `MotorsTab.vue` in this PR. The lines flagged here in `ServosTab.vue` were copy-derived from `MotorsTab.vue`, which already lives in master with the legacy prefix form. Updating `MotorsTab.vue` is unrelated to the Servos-tab resource-pin UI scope of #4881, so it's better left to a small dedicated follow-up that can cover all the other legacy `!utility` usages in the codebase (`SensorGraph.vue`, `LedStripTab.vue`, `ConnectButton.vue`, `LedGrid.vue`) at once.

File: src/components/tabs/ServosTab.vue [hunk 4]
```diff
@@ -169,6 +235,14 @@ const servoConfigs = reactive([]);
 const servoData = reactive([]);
 const originalConfigs = ref("");
 
+// Resource assignment state
+const motorResources = reactive([]);
+const servoResources = reactive([]);
+const hasResourceData = ref(false);
+const resourcesModified = ref(false);
```
[the remark this answers is not in the document]
Comment (Manwe-777) [reply src/components/tabs/ServosTab.vue:242]: Good catch — confirmed bug. `resourcesModified` was being set on every pin change but never participated in `configHasChanged`, so the Save button stayed disabled and `MSP_EEPROM_WRITE` never ran for resource-only edits. `setMotorServoResource` sends the change live to the FC immediately, but without an EEPROM write the assignment is lost on reboot.

Applied the suggested fix in commit `5c915065`:

- `configHasChanged` now ORs in `resourcesModified.value` so any pin edit enables Save.
- The EEPROM-write callback in `updateServos` resets `resourcesModified.value = false` alongside `originalConfigs.value`, so the dirty state clears on successful persist.

Lint + tests still green (52/52).

**Comment by Manwe-777** [comment]
@coderabbitai re: the outside-diff `availablePins` topology suggestion in [#pullrequestreview-4358164974](https://github.com/betaflight/betaflight-configurator/pull/4881#pullrequestreview-4358164974) — the concern is valid but cannot be addressed in this PR because the data source the suggestion needs does not exist on the MSP wire today.

I cross-checked against the companion firmware PR [betaflight/betaflight#14943](https://github.com/betaflight/betaflight/pull/14943), whose `MSP2_MOTOR_SERVO_RESOURCE` handler in `src/main/msp/msp.c` writes exactly:

```
[motorCount][servoCount][motor0_ioTag]…[motorN_ioTag][servo0_ioTag]…[servoN_ioTag]
```

with `motorCount = MAX_SUPPORTED_MOTORS` and `servoCount = MAX_SUPPORTED_SERVOS` (compile-time constants), and `ioTag = 0` for unassigned slots. There is no `MSP2_BOARD_PINS` / `MSP2_TIMER_PINS`, and `MSP_BOARD_INFO` returns only the board identifier — not its timer-capable pad list. The configurator's `MSPHelper.js` reader in this PR matches that exact wire format and consumes nothing more. Without a firmware-side enumeration of the board's timer-capable pads, `availablePins` has no source of truth for free pads to seed a board-derived pool from.

Implementing the suggestion requires, in this order:
1. A new firmware MSP (e.g. `MSP2_BOARD_TIMER_PINS`) that walks `timerHardware[]` and serialises the full candidate-pad list with timer/DMA topology — a separate firmware PR on top of #14943.
2. `MSPHelper.js` plumbing and a new `FC.BOARD_PINS`-style field — a follow-up configurator PR on top of this one.
3. `availablePins` reseeded from the board list, unioned with current assignments.

Until then, the existing `availablePins` correctly covers **swap / revert / unassign-to-NONE** flows by unioning `initialPins` (seed of currently-assigned pads from `MSP2_MOTOR_SERVO_RESOURCE`) with live edits. The **promote-NONE-to-fresh-pin** flow is the known gap, and the existing `servosResourceHint` i18n string already directs users to the CLI `resource MOTOR|SERVO <i> <pin>` command for that case.

You labelled this 🏗️ Heavy lift and posted it as an outside-diff comment yourself, so treating it as tracked follow-up rather than a blocker for this PR.

The two inline review comments you raised earlier ([#3299385709](https://github.com/betaflight/betaflight-configurator/pull/4881#discussion_r3299385709) Tailwind v4 syntax and [#3299385722](https://github.com/betaflight/betaflight-configurator/pull/4881#discussion_r3299385722) `resourcesModified` save-state bug) were both fixed in commit `5c915065` and you've already acknowledged the fixes positively. Could you please run **@coderabbitai full review** against the current HEAD (`5c915065`) and update the review state so the stale Feb `CHANGES_REQUESTED` can be dismissed?

**Review by limonspb — APPROVED** [review]

**Comment by Manwe-777** [comment]
I think this is clear now @haslinghuis @blckmn; not sure why coderabbit's review shows as skipped and requested changes still 

@coderabbitai full review

File: src/components/tabs/ServosTab.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (haslinghuis) [reply src/components/tabs/ServosTab.vue]: @coderabbitai - is this addressed and are then any other remaining issues?

**Comment by haslinghuis** [comment]
Sorry. Really like this - but unfortunate we did not get firmware support merged - updated milestone to be 2026.12

**Review by betaflight-admin — APPROVED** [review]
